app/src/lib/git/branch.ts

```diff
@@ -78,7 +78,13 @@ export async function deleteBranch(
     await deleteLocalBranch(repository, branch.name)
   }
 
-  const remoteName = branch.remote
+  let remoteName = branch.remote
+
+  if (branch.type === BranchType.Remote) {
+    remoteName = branch.name
+      .replace(branch.nameWithoutRemote, '')
+      .replace('/', '')
+  }
 
   if (includeRemote && remoteName) {
```

Comment:
The use of `includeRemote` when you're deleting specifically a remote branch feels odd too (from an API consumer perspective). Like, if I want to delete a remote branch I must call `deleteBranch`, pass the remote branch as a parameter **and** set `includeRemote` to `true`. Shouldn't `includeRemote` be irrelevant once I pass a remote branch? 🤔 

Options I see here:
1. Make `includeRemote` optional, ignore it when the branch is remote, and assume it's `false` when omitted for local branches.
2. Extract the logic to remove remote branches to a `deleteRemoteBranch`, then `deleteBranch` will delete only the local branch or both local and remote when needed, and `DeleteRemoteBranch` would use `deleteRemoteBranch` directly (through the dispatcher, of course, and renaming the previous). More in detail, this is what I'd do for this option:
   1. Inline the implementation of `deleteLocalBranch` in `deleteBranch`.
   2. **BONUS:** remove the `if (branch.type === BranchType.Local)` check since, unless I'm missing something, it wasn't needed before this PR (it was only used with local branches) and it wouldn't be needed after making the changes in this option (2)?
   3. Remove `deleteLocalBranch`
   4. Rename `deleteBranch` to `deleteLocalBranch`. Do the same with `deleteBranch` in the dispatcher and `_deleteBranch` in `AppStore`.
   5. Extract (and reuse) the logic to delete a remote branch to a new `deleteRemoteBranch` function.
   6. Add new `deleteRemoteBranch` in the dispatcher and `_deleteRemoteBranch` in `AppStore` to use this new function from the previous step.

IMO (2) feels much cleaner than (1), but I understand it might be a bit overkill too given the changes we need in the dispatcher and the app store.

What do you think?

Reply:
I like option 2, but kind of like to keep `deleteBranch` existing to be a controller as to determine which of `deleteLocalBranch` and `deleteRemoteBranch` to use.

As for point ii. -> I think there was some defensive programming happening here to prevent the unintentional deletion of remote branches. Because if you had passed in a remote branch to this, it would simply return true and do nothing.

Reply:
> but kind of like to keep `deleteBranch` existing to be a controller as to determine which of `deleteLocalBranch` and `deleteRemoteBranch` to use.

Yeah I considered that too, but thinking about it, it didn't feel like that function had any use with remote branches: its main purpose is to delete local branches, and it would delete remote branches if and only if `includeRemote` is `true`. Technically, it should _always_ delete the local branch (except in completely unexpected situations, hence the defensive programming).

> I think there was some defensive programming happening here

Interesting… I wonder if that check should be switched to a "fast return" and send a non-fatal error to see if it actually happens out in the wild (/cc @niik), and if it doesn't happen just remove the check. In the end `deleteLocalBranch` would only delete the remote branch locally, but that wouldn't impact the remote repo 🤔

Reply:
I looked into doing this.. and the further I got the more I didn't want to do it. :P 

1) If we step back to app-store and look at its `deleteBranch`, It is heavy in promises.. thus splitting it into 2 methods of remote versus local would either be really duplicated functions or passing in a function to call which to me is starting defeat the purpose of being cleaner cause passing functions inside functions that already pass functions hard to read. The `deleteBranch` in branches.ts is very simple and easy to see that it is just a decision maker.

2) `branch-pruner` uses the current deleteLocalBranch.. thus potential to impact it.

Reply:
> 1) If we step back to app-store and look at its `deleteBranch`…

I don't think duplication there would be _that_ bad (in remote branches you don't need to "checkout after delete", because a remote branch can't be the current branch, so the new `_deleteRemoteBranch` method only needs to delete the branch), specially knowing deleting local vs remote branches is a bit different, but I agree all that makes the decision a bit less easy. I'll leave it up to you!

> 2) `branch-pruner` uses the current deleteLocalBranch.. thus potential to impact it.

Nice catch! Seems like inlining that function is not a (good) option then :doh:

However, that wouldn't stop _me_ from renaming `deleteBranch` to `deleteLocalBranch`, and renaming that `deleteLocalBranch` to something more specific like `deleteLocalBranchWithName`?

> The `deleteBranch` in branches.ts is very simple and easy to see that it is just a decision maker.

For me the problem is that you _need_ to look into the implementation/documentation because the function signature is misleading/confusing now 😕 ("what's `includeRemote` if my branch is remote already?"), and I don't think it's hard to make the intent of the function clearer.

Anyway, I personally think the change I'm suggesting would lead to a cleaner result, but I agree it might not be a huge improvement overall vs the amount of changes, so I don't want to block this PR and I'll leave it up to you!

(I'd also love to know what's @niik 's take on this 😆 Sorry!)

Reply:
> (I'd also love to know what's @niik 's take on this 😆 Sorry!)

Hey, sorry about the delay, took me a time to get up to speed here. Great discussion already here :heart: and I think both directions has merit.

What speaks in favor of keeping the deleteBranch method as-is is that I'm a little concerned about the amount of change we're introducing in order to support the branch context menu. While some refactors were inevitable and necessary I wonder if we could have extracted some of the more architectural changes into separate PRs. My gut instinct would have been to try to minimize the surface area where we could introduce bugs as much as possible. I think I would probably just have introduced a new method for deleting a remote branch that could have been used by the new dialog and kept the other changes to a bare minimum to be able to reason about the impact of this change more easily.

On the other hand, I'll say that in general I'm very much in favor of method doing as little as possible (i.e. deleteLocalBranch  and deleteRemoteBranch vs a deleteBranch with a  parameter to include the remote). This goes double for methods in the `git` folder where we should aim for keeping them very focused the git work as opposed to app logic.

So in this case I draw the line at the appStore. In an ideal world (read: this isn't at all the case at the moment and that's fine) app store methods should represent UX actions. I.e. if we show the user a dialog asking them to confirm deletion of a local branch and offer them a choice to also delete the remote branch I think it's absolutely fine for that app store method to take an `includeRemote` parameter as it encapsulates the user's intended operation. That method can then call both `deleteLocalBranch` and `deleteRemoteBranch` which each encapsulate git-level operations.

The presence of a `deleteLocalBranch` in the git layer which optionally deletes the remote branch seems counterintuitive as there's no such operation in git. So my preference would be two dedicated git methods for deleting a local and a remote branch. An added benefit of this is that we no longer need to be passing these methods `Branch` objects as they'll be so much simpler now. The `deleteLocalBranch` can just take a local branch name and the deleteRemoteBranch can take a remote name and a branch to delete on that remote. That way we should get rid of a lot of the trickery of checking whether the branch object it local or not and leave that to the app store method which, if I understand this correctly, is the only caller of these methods where it matters.

I'm like 99% sure I'm just muddying the waters of this discussion by contributing here but those are my initial thoughts. I hope they're helpful but I trust the both of you to find a solution you're both happy with.

Reply:
Thanks for chiming in!! That was a really insightful read ❤️ 

> **So in this case I draw the line at the appStore**. In an ideal world (read: this isn't at all the case at the moment and that's fine) app store methods should represent UX actions. I.e. if we show the user a dialog asking them to confirm deletion of a local branch and offer them a choice to also delete the remote branch I think it's absolutely fine for that app store method to take an includeRemote parameter as it encapsulates the user's intended operation. That method can then call both deleteLocalBranch and deleteRemoteBranch which each encapsulate git-level operations.

I personally like this idea.

> On the other hand, I'll say that in general I'm very much in favor of method doing as little as possible (i.e. deleteLocalBranch and deleteRemoteBranch vs a deleteBranch with a parameter to include the remote). This goes double for methods in the git folder where we should aim for keeping them very focused the git work as opposed to app logic.

And this is a very nice takeaway 😍 

So **ideally** the logic in `deleteBranch` would be moved up to the app store, while in `git` land we'd have the much simpler `deleteLocalBranch` and `deleteRemoteBranch`. I don't think that should be too problematic… except for the fact that we'd lose some valuable tests from `branch-test.ts`.

So I think the best trade-off right now would be to have only one `_deleteBranch` function in app store as you suggest to reflect the UX action, move the current `deleteBranch` from `branch.ts` to its own file (for a clear separation between core `git` stuff vs higher level policies) and be more specific about what it does (`deleteLocalBranchAndUpstream`? `deleteLocalAndTrackingBranch`? `deleteLocalAndRemoteBranch`?), while keeping its tests. Then the app store `_deleteBranch` function will either:
1. Call the shiny new `deleteLocalBranchAndUpstream` (or whatever) for local branches.
2. Call `deleteRemoteBranch` for remote branches.

That said, I agree this might make more sense in a different PR, to keep the refactor part away from the actual feature 😅

Reply:
Awesome discussions! 

I am definitely in favor of moving this refactor out to it's own PR. 

Taking in to consideration @niik's other comments. I think I will create a separate PR to address adding the canonical ref to the branch and making the remoteName function there to provide this functionality separately. 

Additionally, seems the fix needed in the comment about possibly being able to delete non intended remote branches would be good on top of the delete functions refactor PR. 

Going to move this PR back to draft. After those issue get addressed, I will merge into the add-context-menu and go from there.  Thank you both for your great input!

Reply:
I believe this is resolved now